### PR TITLE
Reduce cost of json Unmarshal operations

### PR DESCRIPTION
**Pull request by Jarema**
Until now, in JS API we were checking if request payload is empty.
That check was unmarshaling that payload before we did second unmarshal
for the destination type.
This commit adds separate check to avoid double unmarshaling.

This PR does not change `isEmptyRequest` implementation, as its a less strict check.

Signed-off-by: Tomasz Pietrek <tomasz@nats.io>

### CODE REVIEW

File: server/jetstream_test.go [hunk 1]
```diff
@@ -24365,3 +24365,25 @@ func TestJetStreamStreamConfigClone(t *testing.T) {
...
+		name  string
+		data  []byte
+		valid bool
+	}{
+		{"empty", []byte{}, false},
+		{"empty_object", []byte("{}"), true},
+		{"empty object, not trimmed", []byte("\t\n\r{ }"), true},
+		{"empty_array", []byte("[]"), true},
+		{"empty array, not trimmed", []byte("\t\n\r[ ]"), true},
+		{"empty_string", []byte("\"\""), false},
```
Comment (aricart) [inline server/jetstream_test.go:24380]: Perhaps this is out of scope, but "" is a valid JSON marshal of just an empty string.

**Review by derekcollison — COMMENTED** [review]
Help me understand a bit better what this is doing?

If the len is non-zero it should be unmarshaled. And if that fails should give an error back to the app tier.

Not clear to me the double unmarshal comment.

**Comment by ripienaar** [comment]
the PR is attempting to avoid this unmarshal here 

https://github.com/nats-io/nats-server/blob/ef7d9125ed7bf3678eef482fce63625e06e9c683/server/jetstream_api.go#L2891

**Comment by derekcollison** [comment]
Ah gotcha! Thanks and apologies did not look at the complete function for isEmptyRequest.

**Comment by derekcollison** [comment]
Let's address @aricart comment then should be good to go. Thanks @Jarema !

**Comment by wallyqs** [comment]
There is `json.Valid([]byte)` in the stdlib, that would be lighter than the call to `json.Unmarshal`: https://pkg.go.dev/encoding/json#Valid

**Comment by derekcollison** [comment]
Should we use that instead @Jarema and just change isEmptyRequest?

**Comment by ripienaar** [comment]
The code isn’t checking if it’s valid json. It’s checking if the request - even if valid - is an empty object.

**Comment by Jarema** [comment]
As R.I. mentioned - `json.Unmarshal` in `isEmptyRequest` does not check just for validity.

File: server/jetstream_test.go (hunk 1 above)
Comment (Jarema) [reply server/jetstream_test.go:24381]: This function accepts only JSON objects and JSON arrays, hence the test. Added more tests, renamed function, added better docs. Let me know if it makes sense now.

**Review by derekcollison — APPROVED** [review]
LGTM

**Pull request merged**